docs/design-documents/features/storage/KVStore/KVStore_design.md

```diff
@@ -0,0 +1,460 @@
+    virtual int remove(const char *key);
+ 
+    // Incremental set API
+    virtual int set_start(set_handle_t *handle, const char *key, size_t final_data_size, uint32_t create_flags);
+    virtual int set_add_data(set_handle_t handle, const void *value_data, size_t data_size);
+    virtual int set_finalize(set_handle_t handle);
+ 
+    // Key iterator
+    virtual int iterator_open(iterator_t *it, const char *prefix = NULL);
+    virtual int iterator_next(iterator_t it, char *key, size_t key_size);
+    virtual int iterator_close(iterator_t it);
+}
+```
+
+## Global Key Value interface design
+
+As mentioned above, each KVStore API has a parallel C-style API, used globally with a partition name preceding the key name. 
+
+### Global Key Value APIs
+
+```C
+enum kv_create_flags {
+    KV_WRITE_ONCE_FLAG       = (1 << 0),
+    KV_ENCRYPT_FLAG          = (1 << 1),
+    KV_AUTHENTICATE_FLAG     = (1 << 2),
+    KV_ROLLBACK_PROTECT_FLAG = (1 << 3),
+};
+ 
+static const uint32_t KV_MAX_KEY_LENGTH = 128;
+typedef struct _opaque_set_handle *kv_set_handle_t;
+typedef struct _opaque_key_iterator *kv_iterator_t;
+ 
+typedef struct info {
+    size_t size;
+    uint32_t flags;
+} kv_info_t;
+ 
+// Core API
+int kv_set(const char *full_name_key, const void *buffer, size_t size, uint32_t create_flags);
+int kv_get(const char *full_name_key, void *buffer, size_t buffer_size, size_t *actual_size);
+int kv_get_info(const char *full_name_key, kv_info_t *info);
+int kv_remove(const char *full_name_key);
+ 
+// Key iterator
+int kv_iterator_open(kv_iterator_t *it, const char *full_prefix = nullptr);
+int kv_iterator_next(kv_iterator_t it, char *key, size_t key_size);
+int kv_iterator_close(kv_iterator_t it);
+```
+
+## Mapping APIs
+
+In order to use the global C style APIs, we need APIs to map the partition name to the instance of the implementing KVStore class, typically called once at initialization time. So, for example a "/tdbstore/key1" name means that we wish to access "key1" key in a TDBStore instance. This means that we will need to attach "tdbstore" string to the TDBStore instance at initialization time.  
+These APIs will be part of a different header file ("kv_map.h"), as they serve the integration code and not the KVStore user code.  
+APIs are the following:  
+
+```C
+// Attach and detach
+int kv_init();
+int kv_attach(const char *partition_name, KVStore *kv_instance);
+int kv_detach(const char *partition_name);
+
+// Full name lookup and then break it into KVStore instance and key
+int kv_lookup(const char *full_name, KVStore& *kv_instance, char *key);
+```
+
+## Implementation
+
+Following is the implementation of the Global Key Value interface and of the attachment APIs. KVStore class has no implemetation as it's an interface class.
+
+### Important data structures
+
+```C
+// incremental set handle
+typedef struct {
+    KVStore *kvstore_intance;
+    KVStore::set_handle_t *set_handle;
+} inc_set_handle_t;
```

Comment:
Should these structs have a `kv_` prefix?

Reply:
Good idea. Will change.